skfp/fingerprints/lingo.py

```diff
@@ -0,0 +1,183 @@
+        batch_size: Optional[int] = None,
+        verbose: int = 0,
+    ):
+        super().__init__(
+            n_features_out=fp_size,
+            sparse=sparse,
+            n_jobs=n_jobs,
+            batch_size=batch_size,
+            verbose=verbose,
+        )
+        self.fp_size = fp_size
+        self.substring_length = substring_length
+        self.count = count
+
+    def transform(
+        self, X: Sequence[Union[str, Mol]], copy: bool = False
+    ) -> Union[np.ndarray, csr_array]:
+        """
+        Compute Lingo fingerprints.
+
+        Parameters
+        ----------
+        X : {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        copy : bool, default=False
+            Whether to copy X or modify it in place.
+
+        Returns
+        -------
+        X: {ndarray, sparse matrix} of shape (n_samples, self.fp_size)
+            2D array containing Lingo fingerprints.
+        """
+        return super().transform(X, copy)
+
+    def smiles_to_dicts(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> list[defaultdict[str, int]]:
+        """
+        Convert SMILES strings to dictionaries of substring counts.
+
+        Parameters
+        ----------
+        X: {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        Returns
+        -------
+        result: list[defaultdict[str, int]]
+            List of dictionaries containing substring counts.
+        """
+        X = ensure_smiles(X)
+        X = [re.sub(r"[123456789]", "0", x) for x in X]
+        X = [re.sub(r"Cl", "L", x) for x in X]
+        X = [re.sub(r"Br", "R", x) for x in X]
+
+        result = []
+        for smiles in X:
+            result_dict: defaultdict[str, int] = defaultdict(int)
+            for i in range(len(smiles) - self.substring_length + 1):
+                result_dict[smiles[i : i + self.substring_length]] += 1
+            result.append(result_dict)
+        return result
+
+    def _calculate_fingerprint(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> Union[np.ndarray, csr_array]:
+        X = self.smiles_to_dicts(X)
+        X = self._dicts_to_array(X)
+        return csr_array(X) if self.sparse else X
+
+    def _dicts_to_array(self, X: list[defaultdict[str, int]]) -> np.ndarray:
+        dtype = np.uint32 if self.count else np.uint8
+        result = np.zeros((len(X), self.fp_size), dtype=dtype)
+        for i, dictionary in enumerate(X):
+            for key, value in dictionary.items():
+                string_bytes = key.encode()
+                hash_bytes = hashlib.sha1(string_bytes, usedforsecurity=False).digest()
+                hash_index = int.from_bytes(hash_bytes, byteorder="big") % self.fp_size
+                if self.count:
+                    result[i, hash_index] += value
+                else:
+                    result[i, hash_index] = 1
```

Comment:
First construct the count array, then binarize if needed. if/else in the loop is most likely slower than vectorized NumPy operation.

Reply:
replacing given fragment with:
```             
                # if self.count:
                #     result[i, hash_index] += value
                # else:
                #     result[i, hash_index] = 1
                result[i, hash_index] += value

        # binarize the result
        result[result > 1] = 1
```
actually increased calculation time for hiv dataset from 3.8s to 5.6s. If you have any more specific recommendation, we can use it. My implementation of binarization is significantly slower. Maybe other changes can help in speeding up the processing.